NXmpes: Add concepts first defined in xps sub app-def

lukaspie opened the pull request:
Adding some concepts defined in #30 that @rettigl suggested should be moved to NXmpes

@domna

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -363,7 +397,15 @@ NXmpes(NXobject):
           delay, spin index, temperature, etc. The axes traces should be linked to the
           actual encoder position in NXinstrument or calibrated axes in NXprocess.
-
+      energy(NX_NUMBER):
```
There should be at least one energy axis, either of type binding or kinetic energy. Are there applications of MPES were this is not the case?

domna replied:
PEEM (#65) could potentially be measured without an energy axis.

lukaspie replied:
Hm right, then it cannot be required. Anyway, as discussed yesterday, we need to talk about this further, since at the moment, basically any data can be stored in NXmpes. We want to somehow enumerate which axes are to be expected in order for an experiment to be called MPES.

domna replied:
I mean we can also say that PEEM is not part of NXmpes if it does not require an energy axis (i.e. is it still a photoemission _spectroscopy_ then!?) or we introduce the requirement further down the chain, e.g., building an NXmpes_energy_resolved (not very fond of the naming, though) in between (or above, like NXphotoemission -> NXmpes; NXphotoemission -> NXpeem). We can also give some axis but set them to recommended/optional so users might pick what they need. This would at least define the concepts we are using.

lukaspie replied:
Yeah, if we say we are talking about _spectroscopy_, then PEEM probably doesn't fit nicely. But I agree, the techniques are so similar that maybe an overarching application like NXphotomission could be useful, with NXmpes requiring an energy axis. But the question remains, what data fields do we require so it is an NXphotoemission experiment?

For know, I have made the energy axis recommended.

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -299,6 +310,27 @@ NXmpes(NXobject):
           exists: recommended
           doc: |
             This is the momentum axis to be used for data plotting.
+      energy_correction(NXcalibration):
```
This is probably still too long, but you need all of these fields for such kind of energy correction. The rationale for having this here is that often (at least in XPS), the energy axis is adjusted by shifting all energies uniformally until one peak (or the Fermi edge) is located at a known "correct" energy.

lukaspie commented:
Some things that I would like to add:

- [ ] transmission function of the collection column, see here: https://github.com/FAIRmat-NFDI/nexus_definitions/blob/24c42cd45f2ea466d460c6b20d631b9e057d2313/contributed_definitions/nyaml/NXmpes_xps.yaml#L56 The tranmission function is usually energy vs. relative intensity (could be normalized from 0 to 1). @rettigl preivously suggested that maybe "efficiency" in NXdetector could be a  place to store the transmission function (as NXdata), but it is not just a function of the detector, but mostly of the collection column (depending on the lens mode). It is the response function of the instrument (so maybe related to NXresolution), but can also be used in an intensity calibration.
- [ ] use of flood gun? -> maybe with NXactuator on NXinstrument, then as an environment of NXsample?

lukaspie commented:
> Some things that I would like to add:
> 
> * [ ]  transmission function of the collection column, see here: https://github.com/FAIRmat-NFDI/nexus_definitions/blob/24c42cd45f2ea466d460c6b20d631b9e057d2313/contributed_definitions/nyaml/NXmpes_xps.yaml#L56
>    The transmission function is usually energy vs. relative intensity (could be normalized from 0 to 1). @rettigl previously suggested that maybe "efficiency" in NXdetector could be a  place to store the transmission function (as NXdata), but it is not just a function of the detector, but mostly of the collection column (depending on the lens mode). It is the response function of the instrument (so maybe related to NXresolution), but can also be used in an intensity calibration.

I have added the transmission function as a NXfloat on NXelectronanalyser (in the base class and as recommended in NXmpes). There is also an intensity calibration, which makes use of the transmission function (which could also come from an attached file). However, currently there is no transmission_function in the NXcalibration base_class, so it doesn't quite work yet..

lukaspie commented:
> * [x]  use of flood gun? -> maybe with NXactuator on NXinstrument, then as an environment of NXsample?

Addressed by commit 23c6d4ca806dd9583a65b5f0ff48a75c9a079f35 in #68.

domna reviewed (dismissed):
I just commented on some small formatting and wording issues.

lukaspie commented on contributed_definitions/NXelectronanalyser.nxdl.xml:
```diff
@@ -106,9 +127,16 @@
     </field>
+    <field name="transmission_function" type="NX_FLOAT" units="NX_ANY">
+        <doc>
+             Transmission function of the electron analyzer.
+             
```
Have used something very similar 👍

lukaspie commented on contributed_definitions/NXelectronanalyser.nxdl.xml:
```diff
@@ -106,9 +127,16 @@
             <dim index="1" value="nsa"/>
         </dimensions>
     </field>
+    <field name="transmission_function" type="NX_FLOAT" units="NX_ANY">
```
It should actually be energy vs. intensity (2D), so NX_ENERGY vs NX_UNITLESS, this is why I went with NX_ANY.

domna replied:
ok, wouldn't it be better to use a separate field for the energy scale? 

Or an NXdata for collecting the scale and transmission function, like so:

```
transmission_function(NXdata):
     energy(NX_FLOAT):
         unit: NX_ENERGY
         dim: (n_transmission_function,)
     relative_intensity(NX_FLOAT):
         unit: NX_UNITLESS
         dim: (n_transmission_function,)
```

lukaspie commented on contributed_definitions/NXcalibration.nxdl.xml:
```diff
@@ -108,4 +108,9 @@
     </field>
+    <group name="calibration_file" type="NXnote">
+        <doc>
+             
```
This error occurs if you don't leave one empty line between the yaml and the commented nxdl part (which is added when doing make nyaml).

I have added this empty line again and the doc is generated correctly.

lukaspie commented on contributed_definitions/NXmpes.nxdl.xml:
```diff
@@ -377,6 +406,37 @@ Optional constant settings (like lens focusing parameters on the sample: positio
+                         conductive)?
+                    </doc>
+                </field>
+                <field name="emission_line" recommended="true">
+                    <doc>
+                         Emission line which was used for the calibration
+                    </doc>
+                </field>
```
It would be good to have a concept of an XPS region (or X-ray edge) because this can also be stored on the NXmpes level. See also [here](https://github.com/nexusformat/definitions/pull/1293).

For now, I have just put some examples.

lukaspie replied:
Or rather `NXcore_level` instead of `NXxray_edge`

lukaspie replied:
See also https://github.com/nexusformat/definitions/pull/1293

lukaspie commented:
I have fixed all typos and added some more information in the dosctring where needed.

I did not want to add transmission_function to NXcalibration by name since it is so domain-specific. So, I just called it calibration_input_data in NXcalibration and then implemented it in transmission_correction in NXmpes with explicit naming of the fields. @domna do you think that makes sense?

domna commented:
> I have fixed all typos and added some more information in the dosctring where needed.
> 
> I did not want to add transmission_function to NXcalibration by name since it is so domain-specific. So, I just called it calibration_input_data in NXcalibration and then implemented it in transmission_correction in NXmpes with explicit naming of the fields. @domna do you think that makes sense?

I think it makes totally sense. But I think it may be better to call it differently as this is some kind of transformation/ response function / transfer function of the imaging system, right? So maybe something like `transformation` or `mapping`? This can also apply to other calibrations as these are also transformations or mappings from uncalibrated to calibrated values. What do you think?

domna commented:
> > > I have fixed all typos and added some more information in the dosctring where needed.
> > > I did not want to add transmission_function to NXcalibration by name since it is so domain-specific. So, I just called it calibration_input_data in NXcalibration and then implemented it in transmission_correction in NXmpes with explicit naming of the fields. @domna do you think that makes sense?
> > 
> > 
> > I think it makes totally sense. But I think it may be better to call it differently as this is some kind of transformation/ response function / transfer function of the imaging system, right? So maybe something like `transformation` or `mapping`? This can also apply to other calibrations as these are also transformations or mappings from uncalibrated to calibrated values. What do you think?
> 
> I called it mapping for now. Actually, I put `MAPPING` in all caps, so that I can rename it in the AppDef to `transmission_function`. Not sure if that is the wise thing to do.

This would also allow us to put an unlimited amount of mappings in *one* NXcalibration. Do we want this? I have the feeling we rather want to create multiple NXcalibration where each contains only one mapping. So giving it a fixed name is maybe better even though it might not fit nicely in the appdef level, but we could employ the link-rename scheme we sometimes do (like with the sample and manipulator temperatures, but having two fields `mapping` and `transmission_function` and telling in the appdef level docstring that these are the same concept and should be linked). I know it is not beautiful but I think it might still be better than allowing multiple mappings.

Speaking of this: This is something we could bring in our conceptual discussion next week. Having fields which are the same concept. If we have a way to write this in the appdef the writer could seamlessly check the fields (if both are present) or copy/link the values (if only one is present). This, however, requires to add this as language feature to nxdl.

lukaspie commented:
> > > > I have fixed all typos and added some more information in the dosctring where needed.
> > > > I did not want to add transmission_function to NXcalibration by name since it is so domain-specific. So, I just called it calibration_input_data in NXcalibration and then implemented it in transmission_correction in NXmpes with explicit naming of the fields. @domna do you think that makes sense?
> > > 
> > > 
> > > I think it makes totally sense. But I think it may be better to call it differently as this is some kind of transformation/ response function / transfer function of the imaging system, right? So maybe something like `transformation` or `mapping`? This can also apply to other calibrations as these are also transformations or mappings from uncalibrated to calibrated values. What do you think?
> > 
> > 
> > I called it mapping for now. Actually, I put `MAPPING` in all caps, so that I can rename it in the AppDef to `transmission_function`. Not sure if that is the wise thing to do.
> 
> This would also allow us to put an unlimited amount of mappings in _one_ NXcalibration. Do we want this? I have the feeling we rather want to create multiple NXcalibration where each contains only one mapping. So giving it a fixed name is maybe better even though it might not fit nicely in the appdef level, but we could employ the link-rename scheme we sometimes do (like with the sample and manipulator temperatures, but having two fields `mapping` and `transmission_function` and telling in the appdef level docstring that these are the same concept and should be linked). I know it is not beautiful but I think it might still be better than allowing multiple mappings.
> 
> Speaking of this: This is something we could bring in our conceptual discussion next week. Having fields which are the same concept. If we have a way to write this in the appdef the writer could seamlessly check the fields (if both are present) or copy/link the values (if only one is present). This, however, requires to add this as language feature to nxdl.

I agree that generally such a concept (i.e., that two fields are the same and checking directly) is a good idea. However, in this particular case, the `mapping` function in the calibration can also be read from a file from example and doesn't _need_ to be present on the `electronanalyser` level. So, the checker would need to be smart to only take the data from the link if it actually exists. Anyway, one could even apply a calibration transmission function that is _different_ from the one stored in the analyser (for example, if you suspect that that TF is not correct. 

I have removed the uppercase `MAPPING` in favor of `mapping` for now.

domna commented:
> > > > > I have fixed all typos and added some more information in the dosctring where needed.
> > > > > I did not want to add transmission_function to NXcalibration by name since it is so domain-specific. So, I just called it calibration_input_data in NXcalibration and then implemented it in transmission_correction in NXmpes with explicit naming of the fields. @domna do you think that makes sense?
> > > > 
> > > > 
> > > > I think it makes totally sense. But I think it may be better to call it differently as this is some kind of transformation/ response function / transfer function of the imaging system, right? So maybe something like `transformation` or `mapping`? This can also apply to other calibrations as these are also transformations or mappings from uncalibrated to calibrated values. What do you think?
> > > 
> > > 
> > > I called it mapping for now. Actually, I put `MAPPING` in all caps, so that I can rename it in the AppDef to `transmission_function`. Not sure if that is the wise thing to do.
> > 
> > 
> > This would also allow us to put an unlimited amount of mappings in _one_ NXcalibration. Do we want this? I have the feeling we rather want to create multiple NXcalibration where each contains only one mapping. So giving it a fixed name is maybe better even though it might not fit nicely in the appdef level, but we could employ the link-rename scheme we sometimes do (like with the sample and manipulator temperatures, but having two fields `mapping` and `transmission_function` and telling in the appdef level docstring that these are the same concept and should be linked). I know it is not beautiful but I think it might still be better than allowing multiple mappings.
> > Speaking of this: This is something we could bring in our conceptual discussion next week. Having fields which are the same concept. If we have a way to write this in the appdef the writer could seamlessly check the fields (if both are present) or copy/link the values (if only one is present). This, however, requires to add this as language feature to nxdl.
> 
> I agree that generally such a concept (i.e., that two fields are the same and checking directly) is a good idea. However, in this particular case, the `mapping` function in the calibration can also be read from a file from example and doesn't _need_ to be present on the `electronanalyser` level. So, the checker would need to be smart to only take the data from the link if it actually exists. Anyway, one could even apply a calibration transmission function that is _different_ from the one stored in the analyser (for example, if you suspect that that TF is not correct.
> 
> I have removed the uppercase `MAPPING` in favor of `mapping` for now.

Yes it doesn't have to be present. My idea would be to have `(NXmpes)NXcalibration/transmission_function` and `NXcalibration/mapping`. And we would just state `(NXmpes)NXcalibration/transmission_function` `isSameAs` `NXcalibration/mapping` which would also propagate its optionality. So if we set `transmission_function` to required it would mean that `NXcalibration/mapping` would also be required and the same for optional/recommended.
The linker would need to figure out which one exists and do the linking appropriately or just do nothing when the field optional/recommended and *both* are not present or fail in the case it is required.

For the case for two different transmission functions I would rather say that one should just put two NXcalibrations in the file and explain which was used to transform any data, because I think NXcalibration does carry important metadata which always should belong to the different mappings. One could then also make comparing NXdata's with the two NXcalibrations having different outputs.

lukaspie commented:
> > > > > > I have fixed all typos and added some more information in the dosctring where needed.
> > > > > > I did not want to add transmission_function to NXcalibration by name since it is so domain-specific. So, I just called it calibration_input_data in NXcalibration and then implemented it in transmission_correction in NXmpes with explicit naming of the fields. @domna do you think that makes sense?
> > > > > 
> > > > > 
> > > > > I think it makes totally sense. But I think it may be better to call it differently as this is some kind of transformation/ response function / transfer function of the imaging system, right? So maybe something like `transformation` or `mapping`? This can also apply to other calibrations as these are also transformations or mappings from uncalibrated to calibrated values. What do you think?
> > > > 
> > > > 
> > > > I called it mapping for now. Actually, I put `MAPPING` in all caps, so that I can rename it in the AppDef to `transmission_function`. Not sure if that is the wise thing to do.
> > > 
> > > 
> > > This would also allow us to put an unlimited amount of mappings in _one_ NXcalibration. Do we want this? I have the feeling we rather want to create multiple NXcalibration where each contains only one mapping. So giving it a fixed name is maybe better even though it might not fit nicely in the appdef level, but we could employ the link-rename scheme we sometimes do (like with the sample and manipulator temperatures, but having two fields `mapping` and `transmission_function` and telling in the appdef level docstring that these are the same concept and should be linked). I know it is not beautiful but I think it might still be better than allowing multiple mappings.
> > > Speaking of this: This is something we could bring in our conceptual discussion next week. Having fields which are the same concept. If we have a way to write this in the appdef the writer could seamlessly check the fields (if both are present) or copy/link the values (if only one is present). This, however, requires to add this as language feature to nxdl.
> > 
> > 
> > I agree that generally such a concept (i.e., that two fields are the same and checking directly) is a good idea. However, in this particular case, the `mapping` function in the calibration can also be read from a file from example and doesn't _need_ to be present on the `electronanalyser` level. So, the checker would need to be smart to only take the data from the link if it actually exists. Anyway, one could even apply a calibration transmission function that is _different_ from the one stored in the analyser (for example, if you suspect that that TF is not correct.
> > I have removed the uppercase `MAPPING` in favor of `mapping` for now.
> 
> Yes it doesn't have to be present. My idea would be to have `(NXmpes)NXcalibration/transmission_function` and `NXcalibration/mapping`. And we would just state `(NXmpes)NXcalibration/transmission_function` `isSameAs` `NXcalibration/mapping` which would also propagate its optionality. So if we set `transmission_function` to required it would mean that `NXcalibration/mapping` would also be required and the same for optional/recommended. The linker would need to figure out which one exists and do the linking appropriately or just do nothing when the field optional/recommended and _both_ are not present or fail in the case it is required.
> 
> For the case for two different transmission functions I would rather say that one should just put two NXcalibrations in the file and explain which was used to transform any data, because I think NXcalibration does carry important metadata which always should belong to the different mappings. One could then also make comparing NXdata's with the two NXcalibrations having different outputs.

Ok, so the structure would be the following?

Nxmpes.yaml
```
(NXprocess):
  transmission_correction(NXcalibration):
    transmission_function(NXdata):
      exists: recommended
      isSameAs: Nxcalibration/mapping
      ## here should be the identifier that this is the same concept as _mapping_ defined in NXcalibration
```

Nxcalibration.yaml
```
(NXprocess):
  mapping(NXdata):
```

Since we don't have this linker as of yet, I have kept the naming in both NXmpes and NXcalibration the same for now. Other than that, I think we can merge this branch back to `mpes-refactor`.

domna commented:
> > > > > > > I have fixed all typos and added some more information in the dosctring where needed.
> > > > > > > I did not want to add transmission_function to NXcalibration by name since it is so domain-specific. So, I just called it calibration_input_data in NXcalibration and then implemented it in transmission_correction in NXmpes with explicit naming of the fields. @domna do you think that makes sense?
> > > > > > 
> > > > > > 
> > > > > > I think it makes totally sense. But I think it may be better to call it differently as this is some kind of transformation/ response function / transfer function of the imaging system, right? So maybe something like `transformation` or `mapping`? This can also apply to other calibrations as these are also transformations or mappings from uncalibrated to calibrated values. What do you think?
> > > > > 
> > > > > 
> > > > > I called it mapping for now. Actually, I put `MAPPING` in all caps, so that I can rename it in the AppDef to `transmission_function`. Not sure if that is the wise thing to do.
> > > > 
> > > > 
> > > > This would also allow us to put an unlimited amount of mappings in _one_ NXcalibration. Do we want this? I have the feeling we rather want to create multiple NXcalibration where each contains only one mapping. So giving it a fixed name is maybe better even though it might not fit nicely in the appdef level, but we could employ the link-rename scheme we sometimes do (like with the sample and manipulator temperatures, but having two fields `mapping` and `transmission_function` and telling in the appdef level docstring that these are the same concept and should be linked). I know it is not beautiful but I think it might still be better than allowing multiple mappings.
> > > > Speaking of this: This is something we could bring in our conceptual discussion next week. Having fields which are the same concept. If we have a way to write this in the appdef the writer could seamlessly check the fields (if both are present) or copy/link the values (if only one is present). This, however, requires to add this as language feature to nxdl.
> > > 
> > > 
> > > I agree that generally such a concept (i.e., that two fields are the same and checking directly) is a good idea. However, in this particular case, the `mapping` function in the calibration can also be read from a file from example and doesn't _need_ to be present on the `electronanalyser` level. So, the checker would need to be smart to only take the data from the link if it actually exists. Anyway, one could even apply a calibration transmission function that is _different_ from the one stored in the analyser (for example, if you suspect that that TF is not correct.
> > > I have removed the uppercase `MAPPING` in favor of `mapping` for now.
> > 
> > 
> > Yes it doesn't have to be present. My idea would be to have `(NXmpes)NXcalibration/transmission_function` and `NXcalibration/mapping`. And we would just state `(NXmpes)NXcalibration/transmission_function` `isSameAs` `NXcalibration/mapping` which would also propagate its optionality. So if we set `transmission_function` to required it would mean that `NXcalibration/mapping` would also be required and the same for optional/recommended. The linker would need to figure out which one exists and do the linking appropriately or just do nothing when the field optional/recommended and _both_ are not present or fail in the case it is required.
> > For the case for two different transmission functions I would rather say that one should just put two NXcalibrations in the file and explain which was used to transform any data, because I think NXcalibration does carry important metadata which always should belong to the different mappings. One could then also make comparing NXdata's with the two NXcalibrations having different outputs.
> 
> Ok, so the structure would be the following?
> 
> Nxmpes.yaml
> 
> ```
> (NXprocess):
>   transmission_correction(NXcalibration):
>     transmission_function(NXdata):
>       exists: recommended
>       isSameAs: Nxcalibration/mapping
>       ## here should be the identifier that this is the same concept as _mapping_ defined in NXcalibration
> ```
> 
> Nxcalibration.yaml
> 
> ```
> (NXprocess):
>   mapping(NXdata):
> ```
> 

Yes, that's exactly how I had it in mind (using an `(NXcalibration)` instead of `(NXprocess)` in the last snippet though - but I'm sure that was a typo).

> Since we don't have this linker as of yet, I have kept the naming in both NXmpes and NXcalibration the same for now. Other than that, I think we can merge this branch back to `mpes-refactor`.

Perfect, I'll give it a final review and then let's merge it 🎉

domna reviewed (approved):
Looks good. Thanks!
